util/rwlock/src/rwlock.c

```diff
@@ -0,0 +1,213 @@
+    RWLOCK_DBG_ASSERT(lock->handoffs > 0);
+    lock->handoffs--;
+}
+
+/**
+ * Indicates whether a prospective reader must wait for the lock to become
+ * available.
+ */
+static bool
+rwlock_read_must_block(const struct rwlock *lock)
+{
+    RWLOCK_DBG_ASSERT(lock->mtx.mu_owner == g_current_task);
+
+    return lock->active_writer ||
+           lock->pending_writers > 0 ||
+           lock->handoffs > 0;
+}
+
+/**
+ * Indicates whether a prospective writer must wait for the lock to become
+ * available.
+ */
+static bool
+rwlock_write_must_block(const struct rwlock *lock)
+{
+    RWLOCK_DBG_ASSERT(lock->mtx.mu_owner == g_current_task);
+
+    return lock->active_writer ||
+           lock->num_readers > 0 ||
+           lock->handoffs > 0;
+}
+
+void
+rwlock_acquire_read(struct rwlock *lock)
+{
+    bool acquired;
+
+    os_mutex_pend(&lock->mtx, OS_TIMEOUT_NEVER);
+
+    if (rwlock_read_must_block(lock)) {
+        lock->pending_readers++;
+        acquired = false;
+    } else {
+        lock->num_readers++;
+        acquired = true;
+    }
+
+    os_mutex_release(&lock->mtx);
+
+    if (acquired) {
+        /* No contention; lock acquired. */
+        return;
+    }
+
+    /* Wait for the lock to become available. */
+    os_sem_pend(&lock->rsem, OS_TIMEOUT_NEVER);
+
+    /* Record reader ownership. */
+    os_mutex_pend(&lock->mtx, OS_TIMEOUT_NEVER);
+    lock->num_readers++;
+    rwlock_complete_handoff(lock);
+    os_mutex_release(&lock->mtx);
+}
+
+void
+rwlock_release_read(struct rwlock *lock)
+{
+    os_mutex_pend(&lock->mtx, OS_TIMEOUT_NEVER);
+
+    RWLOCK_DBG_ASSERT(lock->num_readers > 0);
+    lock->num_readers--;
+
+    /* If this is the last active reader, unblock a pending writer if there is
+     * one.
+     */
+    if (lock->num_readers == 0) {
+        rwlock_unblock(lock);
+    }
+
+    os_mutex_release(&lock->mtx);
+}
+
+void
+rwlock_acquire_write(struct rwlock *lock)
+{
+    bool acquired;
+
+    os_mutex_pend(&lock->mtx, OS_TIMEOUT_NEVER);
+
+    if (rwlock_write_must_block(lock)) {
+        lock->pending_writers++;
+        acquired = false;
+    } else {
+        lock->active_writer = true;
+        acquired = true;
+    }
+
+    os_mutex_release(&lock->mtx);
+
+    if (acquired) {
+        /* No contention; lock acquired. */
+        return;
+    }
+
+    /* Wait for the lock to become available. */
+    os_sem_pend(&lock->wsem, OS_TIMEOUT_NEVER);
+
+    /* Record writer ownership. */
+    os_mutex_pend(&lock->mtx, OS_TIMEOUT_NEVER);
+    lock->active_writer = true;
+    rwlock_complete_handoff(lock);
+    os_mutex_release(&lock->mtx);
+}
+
+void
+rwlock_release_write(struct rwlock *lock)
+{
+    os_mutex_pend(&lock->mtx, OS_TIMEOUT_NEVER);
+
+    RWLOCK_DBG_ASSERT(lock->active_writer);
+    lock->active_writer = false;
+
+    rwlock_unblock(lock);
+
+    os_mutex_release(&lock->mtx);
+}
+
+int
+rwlock_init(struct rwlock *lock)
+{
+    int rc;
+
+    *lock = (struct rwlock) { 0 };
```

Comment:
LLVM does not like this kind of initializations.

Reply:
Then LLVM is broken :).  It works in my clang (`Apple LLVM version 7.3.0 (clang-703.0.31)`).  Out of curiosity, which version of clang are you using?

Reply:
Just upgraded, and ended up with Apple LLVM version 9.1.0 (clang-902.0.39.2). Upgrade broke my gcc-5 installation, so I tried running with clang for a bit. A lot of issues when building unit tests. Some of them due to this kind of initialization; so it is fresh in my mind :)
In the end, I ended up falling back to gcc :)

Reply:
Wow, that would be quite a regression for Apple to introduce.  I have an equally recent version of clang on my linux machine (5.0.2) which doesn't have this bug.

I wonder if it is correct to change code to accommodate a broken version of a compiler...